api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java

```diff
@@ -0,0 +1,191 @@
+
+  private static final String AUTH_CODE = "code";
+  private static final String REDIRECT_URL = "url";
+  private static final String ACCESS_TOKEN = "access_token_1";
+  private static final String LOGIN_GOV_USERNAME = "foo@gmail.com";
+  private static final String USER_INFO_JSON_LOGIN_GOV =
+      "{\"preferred_username\":\"user1@Login.Gov.com\", \"email\":\"" + LOGIN_GOV_USERNAME + "\"}";
+  private static final String USER_INFO_JSON_ERA =
+      "{\"preferred_username\":\"user2@eraCommons.com\", \"email\":\"" + LOGIN_GOV_USERNAME + "\"}";
+  private static final String ID_TOKEN_JWT_IAL_1 =
+      JWT.create()
+          .withClaim(
+              ACR_CLAIM,
+              "https://stsstg.nih.gov/assurance/ial/1 https://stsstg.nih.gov/assurance/aal/1")
+          .sign(Algorithm.none());
+  private static final String ID_TOKEN_JWT_IAL_2 =
+      JWT.create()
+          .withClaim(
+              ACR_CLAIM,
+              "https://stsstg.nih.gov/assurance/ial/2 https://stsstg.nih.gov/assurance/aal/1")
+          .sign(Algorithm.none());
+  private static final TokenResponse TOKEN_RESPONSE_IAL1 =
+      new TokenResponse().setAccessToken(ACCESS_TOKEN).set(Id_TOKEN_FIELD_NAME, ID_TOKEN_JWT_IAL_1);
+  private static final TokenResponse TOKEN_RESPONSE_IAL2 =
+      new TokenResponse().setAccessToken(ACCESS_TOKEN).set(Id_TOKEN_FIELD_NAME, ID_TOKEN_JWT_IAL_2);
+
+  private long userId;
+  private static DbUser currentUser;
+
+  private RasLinkService rasLinkService;
+  private ObjectMapper objectMapper = new ObjectMapper();
+
+  @Autowired private UserService userService;
+  @Autowired private UserDao userDao;
+  @Mock private static OpenIdConnectClient mockOidcClient;
+  @Mock private static Provider<OpenIdConnectClient> mockOidcClientProvider;
+  @Mock private static HttpTransport mockHttpTransport;
+  @Mock private OpenIdConnectClient mockRasOidcClient;
+
+  @TestConfiguration
+  @Import({
+    RasLinkService.class,
+    UserServiceTestConfiguration.class,
+  })
+  @MockBean({
+    FireCloudService.class,
+    ComplianceService.class,
+    DirectoryService.class,
+    UserServiceAuditor.class,
+    FreeTierBillingService.class,
+  })
+  static class Configuration {
+    @Bean
+    Clock clock() {
+      return CLOCK;
+    }
+
+    @Bean
+    Random random() {
+      return new FakeLongRandom(123);
+    }
+
+    @Bean
+    HttpTransport httpTransport() {
+      return mockHttpTransport;
+    }
+
+    @Bean
+    @Qualifier(RAS_OIDC_CLIENT)
+    Provider<OpenIdConnectClient> rasOidcClientProvider() {
+      return mockOidcClientProvider;
+    }
+
+    @Bean
+    @Qualifier(RAS_OIDC_CLIENT)
+    OpenIdConnectClient rasOidcClient() {
+      return mockOidcClient;
+    }
+
+    @Bean
+    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    public WorkbenchConfig getWorkbenchConfig() {
+      return WorkbenchConfig.createEmptyConfig();
+    }
+
+    @Bean
+    @Scope("prototype")
+    DbUser user() {
+      return currentUser;
+    }
+  }
+
```

Comment:
If you use the pattern of:

```
@MockBean{
 ...
 OpenIdConnectClient.class
}
class TestConfiguration { 
...
}

@Autowired OpenIdConnectClient mockOidcClient;
```

That should function as a mock. The test config generates the mock, then you autowire (inject) it into your test class.

Reply:
I tried that but I don't know how to put @Qualifier inside mockbean array. So I end up with doing the current way.

Reply:
Ahhh, OK, this makes sense then - I don't know off the top of my head a better pattern